cas_object/src/cas_object_format.rs

```diff
@@ -1303,6 +1304,130 @@ impl CasObject {
+        xorb: RawXorbData,
+        compression_scheme: Option<CompressionScheme>,
+        verify_against_previous: bool,
+    ) -> Result<Self, CasObjectError> {
+        let mut cas = CasObject::default();
+        cas.info.cashash = xorb.hash();
+
+        let xorb_hash = cas.info.cashash;
+        let raw_num_bytes = xorb.num_bytes() as u64;
+        let num_chunks = xorb.data.len();
+
+        let chunks_and_boundaries = xorb.cas_info.chunks_and_boundaries();
+
+        // For testing, check the current serialization method against the older version, byte for byte.
+        let mut verification_data = Vec::new();
+
+        if verify_against_previous {
+            // This is the older path;
+            let xorb_data = xorb.to_vec();
+
+            let mut writer = Cursor::new(&mut verification_data);
+
+            let (_, nbytes_trans) =
+                CasObject::serialize(&mut writer, &xorb_hash, &xorb_data, &chunks_and_boundaries, compression_scheme)?;
+
+            assert_eq!(verification_data.len(), nbytes_trans);
+        }
+
+        cas.info.num_chunks = chunks_and_boundaries.len() as u32;
+        cas.info.chunk_boundary_offsets = Vec::with_capacity(cas.info.num_chunks as usize);
+        cas.info.chunk_hashes = chunks_and_boundaries.iter().map(|(hash, _)| *hash).collect();
+        cas.info.unpacked_chunk_offsets = chunks_and_boundaries
+            .iter()
+            .map(|(_, unpacked_chunk_boundary)| *unpacked_chunk_boundary)
+            .collect();
+
+        // Make an upper estimate of the size to account for to avoid copies and allocations. We assume compression
+        // doesn't help at all, and add just a bit of buffer on it (lz4 with no compression adds just a few
+        // bytes).
+        let size_upper_bound = xorb.num_bytes()
+            + size_of::<CasObjectInfoV1>()
+            + (32 + 2 * size_of::<u32>() + size_of::<MerkleHash>() + size_of::<CASChunkHeader>()) * xorb.data.len();
+
+        let mut serialized_data = Vec::with_capacity(size_upper_bound);
+
+        {
+            let mut writer = Cursor::new(&mut serialized_data);
+
+            for chunk in xorb.data {
+                // now serialize chunk directly to writer (since chunks come first!)
+                serialize_chunk(&chunk, &mut writer, compression_scheme)?;
+                cas.info.chunk_boundary_offsets.push(writer.stream_position()? as u32);
+            }
+
+            // Fill in the boundary offsets
+            cas.info.fill_in_boundary_offsets();
+
+            // now that footer is ready, write out to writer.
+            let info_length = cas.info.serialize(&mut writer)?;
+            cas.info_length = info_length as u32;
+
+            writer.write_all(&cas.info_length.to_le_bytes())?;
+        }
+
+        Ok(Self {
+            serialized_data,
+            hash: xorb_hash,
+            raw_num_bytes,
+            num_chunks,
+        })
+    }
+
+    // Used in testing
+    #[allow(dead_code)]
```

Comment:
I think if this is just needed for testing, you can probably just put it under the `test_utils` module below (also, we should probably tag that module with a `#[cfg(test)]`)?